File: src/typescript/Array-String/String/LongestSubstringWindow.ts
// 76. Minimum Window Substring https://leetcode.com/problems/minimum-window-substring/
function minWindow(s: string, t: string): string {
    if (t == "")
        return "";

    let tMap: any = {}, sMap: any = {};
    let need: number = 0, have: number = 0, left: number = 0, right: number = 0;
    let res: string = "";

    for (let i = 0; i < t.length; i++) {
        if (tMap[t[i]]) {
            tMap[t[i]]++;
        } else {
            tMap[t[i]] = 1;
            need++;
        }
    }

    while (right <= s.length && left <= (s.length - t.length)) {
        if (have >= need) {
            let t1 = tMap[s[left]];
            if (res.length > s.substring(left, right).length || res.length === 0)
                res = s.substring(left, right);

            if (t1) {
                sMap[s[left]]--;
                if (sMap[s[left]] < t1)
                    have--;
            }
            left++;
        } else {
            let t1 = tMap[s[right]];
            let s1 = sMap[s[right]];

            if (t1 && s1)
                sMap[s[right]]++;
            else if (t1 && !s1)
                sMap[s[right]] = 1;

            if (t1 && t1 === sMap[s[right]])
                have++;

            right++;
        }
    }

    return res;
};

console.log(minWindow("a", "a"));

File: src/Array n String/String/LongestSubStringWindow.cpp
// 76. Minimum Window Substring https://leetcode.com/problems/minimum-window-substring/

#include <iostream>
#include <unordered_map>
#include <string>

using namespace std;

class Solution
{
public:
	string minWindow(string s, string t)
	{
		unordered_map<char, int> t_map, s_map;
		int l = 0, r = 0, need = 0, have = 0;
		string res = "";

		for (int i = 0; i < s.length(); i++)
		{
			if (t_map[s[i]])
			{
				t_map[s[i]]++;
			}
			else
			{
				t_map[s[i]] = 1;
				need++;
			}
		}

		while (r < s.length() && l < (s.length() - t.length()))
		{
			if (have >= need)
			{
				l++;
			}
			else
			{
				r++;
				if (t_map[s[r]])
					;
			}
		}

		return res;
	}
};

using namespace std;

int main()
{
	Solution s;
	cout << s.minWindow("Mayanak kumar", "Mayank");
}
